src/new-storage/BatchManager.js

```diff
@@ -0,0 +1,161 @@
+        }
+
+        this.opts = {
+            ...defaultOptions,
+            ...opts
+        }
+
+        // bucketId => batch
+        this.batches = {}
+        // batchId => batch
+        this.pendingBatches = {}
+
+        this.cassandraClient = cassandraClient
+        this.insertStatement = this.opts.useTtl ? INSERT_STATEMENT_WITH_TTL : INSERT_STATEMENT
+    }
+
+    store(bucketId, streamMessage) {
+        const batch = this.batches[bucketId]
+
+        if (batch && batch.isFull()) {
+            batch.setClose(false)
+            this._moveFullBatch(bucketId, batch)
+        }
+
+        if (this.batches[bucketId] === undefined) {
+            debug('creating new batch')
+            const newBatch = new Batch(bucketId, this.opts.batchMaxSize, this.opts.batchMaxRecords, this.opts.batchCloseTimeout, this.opts.batchMaxRetries)
+            newBatch.on('state', (batchBucketId, id, state, size, numberOfRecords) => this._batchChangedState(batchBucketId, id, state, size, numberOfRecords))
+            this.batches[bucketId] = newBatch
+        }
+
+        this.batches[bucketId].push(streamMessage)
+    }
+
+    _moveFullBatch(bucketId, batch) {
+        debug('moving batch to pendingBatches')
+        this.pendingBatches[batch.getId()] = batch
+        this.pendingBatches[batch.getId()].scheduleInsert()
+        delete this.batches[bucketId]
+    }
+
+    _batchChangedState(bucketId, batchId, state, size, numberOfRecords) {
+        debug(`_batchChangedState, bucketId: ${bucketId}, id: ${batchId}, state: ${state}, size: ${size}, records: ${numberOfRecords}`)
+        if (state === Batch.states.PENDING) {
+            this._insert(batchId)
+        } else if (state === Batch.states.CLOSED) {
+            const batch = this.batches[bucketId]
+            if (batch) {
+                this._moveFullBatch(bucketId, batch)
+            }
+        }
+    }
+
+    stop() {
+        Object.values(this.batches).forEach((batch) => batch.clear())
+        Object.values(this.pendingBatches).forEach((batch) => batch.clear())
+    }
+
+    async _insert(batchId) {
+        const batch = this.pendingBatches[batchId]
+
+        try {
+            const queries = batch.streamMessages.map((streamMessage) => {
+                return {
+                    query: this.insertStatement,
+                    params: [
+                        streamMessage.getStreamId(),
+                        streamMessage.getStreamPartition(),
+                        // TODO to be added in next PR batch.getBucketId(),
+                        streamMessage.getTimestamp(),
+                        streamMessage.getSequenceNumber(),
+                        streamMessage.getPublisherId(),
+                        streamMessage.getMsgChainId(),
+                        Buffer.from(streamMessage.serialize()),
+                    ]
+                }
+            })
+
+            await this.cassandraClient.batch(queries, {
```

Comment:
No, because we come to this code by timeout `newBatch.on('pending', () => this._insert(newBatch.getId()))`
And then we if failed set `scheduleInsert` again

Reply:
Ah indeed you're right. I confused `setTimeout` with `setInterval`. LGTM.